### PR TITLE
[TRTLLM-9767][feat] Enable attention dp for qwen3-next.

### PR DESCRIPTION
**Pull request by nv-guomingz**
### Prerequisite
- [x] Depends on #9873 

## Summary by CodeRabbit

* **New Features**
  * Added support for Disaggregated Serving in Qwen3Next model.
  * Introduced attention data parallelism (attention_dp) as a new configurable option for enhanced model parallelization.
  * Improved Mamba cache handling with padding-aware state reordering.

* **Bug Fixes**
  * Optimized weight padding logic for more efficient memory operations.

* **Tests**
  * Expanded test coverage for attention data parallelism configurations.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by kmk142789 — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 1]
```diff
@@ -192,27 +202,32 @@ def forward(
         use_dp_padding = False
         all_rank_num_tokens = attn_metadata.all_rank_num_tokens
 
+        if self.enable_attention_dp and self.mapping.tp_size > 1 and get_sm_version(
+        ) == 120:
+            use_dp_padding = True
+            hidden_states = torch.nn.functional.pad(
+                hidden_states,
+                (0, 0, 0, max(all_rank_num_tokens) - hidden_states.shape[0]))
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py]: I am curious why we need to use this padding, can we add a comment for it?

**Review by Wanli-Jiang — APPROVED** [review]

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 2]
```diff
@@ -4875,19 +4880,26 @@ def test_bf16_4gpu(self, tp_size, pp_size, ep_size, cuda_graph,
...
-                   overlap_scheduler, mocker):
+                   overlap_scheduler, attention_dp, mocker):
         model_path = f"{self.MODEL_PATH}/qwen3-next-80b-instruct-nvfp4-ptq-fp8kv"
 
         kv_cache_config = KvCacheConfig(free_gpu_memory_fraction=0.6,
                                         enable_block_reuse=False)
         pytorch_config = dict(disable_overlap_scheduler=not overlap_scheduler,
                               cuda_graph_config=CudaGraphConfig(
-                                  max_batch_size=512, enable_padding=True)
+                                  max_batch_size=512, enable_padding=False)
```
Comment (yuxianq) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py]: Why do we set `enable_padding` to False here?

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 3]
```diff
@@ -517,7 +531,8 @@ def __init__(self,
             mapping=mapping,
             tensor_parallel_mode=TensorParallelMode.ROW,
             quant_config=model_config.get_quant_config(),
-            reduce_output=True,
+            reduce_output=not self.model_config.mapping.enable_attention_dp
```
Comment (yuxianq) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Same suggestion, we can update `reduce_output` inside `Linear` instead.

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 4]
```diff
@@ -865,8 +879,13 @@ def __init__(
                                                 use_gemma=True)
         self.layer_idx = layer_idx
 
-        self.allreduce = AllReduce(mapping=model_config.mapping,
-                                   strategy=model_config.allreduce_strategy)
+        # Only create allreduce when not using attention DP
+        if not self.enable_attention_dp and model_config.mapping.tp_size > 1:
+            self.allreduce = AllReduce(mapping=model_config.mapping,
```
Comment (yuxianq) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Same suggestion, we can always create `AllReduce`.

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 5]
```diff
@@ -1002,13 +1020,16 @@ def __init__(self, model_config: ModelConfig[Qwen3NextConfig],
...
         config = model_config.pretrained_config
 
+        self.mapping = model_config.mapping
+        self.enable_attention_dp = self.mapping.enable_attention_dp
+
         self.self_attn = Qwen3NextAttention(
             model_config,
             layer_idx=layer_idx,
             fuse_qk_norm_rope=False,
+            reduce_output=not self.enable_attention_dp
```
Comment (yuxianq) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py]: If we update `reduce_output` inside `Linear`, we don't need to thread this `reduce_output` through `Qwen3NextAttention->Qwen3Attention->QKNormRoPEAttention-> Attention->Linear`.

Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Updated

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 6]
```diff
@@ -1025,23 +1046,27 @@ def __init__(self, model_config: ModelConfig[Qwen3NextConfig],
                                                 use_gemma=True)
         self.layer_idx = layer_idx
 
-        self.allreduce = AllReduce(mapping=model_config.mapping,
-                                   strategy=model_config.allreduce_strategy)
+        # Only create allreduce when not using attention DP
+        if not self.enable_attention_dp and model_config.mapping.tp_size > 1:
+            self.allreduce = AllReduce(mapping=model_config.mapping,
```
[the remark this answers is not in the document]
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Updated

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 4 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Updated

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 7]
```diff
@@ -165,14 +171,18 @@ def __init__(
             bias=config.mlp_bias if hasattr(config, 'mlp_bias') else False,
             dtype=config.torch_dtype,
             config=model_config,
-            reduce_output=False,
-        )
+            reduce_output=self.enable_attention_dp and self.mapping.tp_size > 1,
```
[the remark this answers is not in the document]
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Updated in Linear class

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 8]
```diff
@@ -133,8 +133,11 @@ def __init__(
         self.top_k = config.num_experts_per_tok
         self.enable_attention_dp = model_config.mapping.enable_attention_dp
         self.mapping = model_config.mapping
+
+        # Only create allreduce when not using attention DP
```
Comment (yuxianq) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Remove this comment

**Review by yuxianq — APPROVED** [review]

**Review by jieli-matrix — APPROVED** [review]

**Review by Shixiaowei02 — APPROVED** [review]
